### PR TITLE
Add `timedWait` to `std.Thread.Semaphore`

### PR DESCRIPTION
**Pull request by fkollmann**
Changes:

- Add `timedWait` to `std.Thread.Semaphore`
- Add example to documentation of `std.Thread.Semaphore`

Motivation:

The `std.Thread.Condition` already supports `timedWait` and this brings the functionality to the same level.

Also, this allows for easy consumer thread waits:

```zig
fn consumerThreadMain() {
  while (!stop) {
    sem.timedWait(25 * std.time.ns_per_ms) catch continue;
  }
}
```

### CODE REVIEW

File: lib/std/Thread/Semaphore.zig [hunk 1]
```diff
@@ -26,6 +43,18 @@ pub fn wait(sem: *Semaphore) void {
         sem.cond.signal();
 }
 
+pub fn timedWait(sem: *Semaphore, timeout_ns: u64) error{Timeout}!void {
+    sem.mutex.lock();
+    defer sem.mutex.unlock();
+
+    while (sem.permits == 0)
+        sem.cond.timedWait(&sem.mutex, timeout_ns);
```
Comment (rootbeer) [inline lib/std/Thread/Semaphore.zig:51]: Is this line missing a `try` on the call to timedWait?

**Review by rootbeer — COMMENTED** [review]
Should probably add a call to `timedWait` in the test case at the end of the file (or add a new test case).  Ideally, one call with a very short timeout that will expire, and one call with a very long timeout that will resolve.  (I'm not a Zig team member, so compliance is no guarantee of submission, etc.)

**Comment by fkollmann** [comment]
Took me some time to get the build running...

I added a unit test which successfully completed by running `build/stage3/bin/zig build test-std -Dskip-release -Dskip-non-native -Dtest-filter="Thread.Semaphore"`

(I also did a negative test, by having the unit test to return an error to check, if the test was actually done, or not.)

**Comment by fkollmann** [comment]
Please let me know if you would prefer a black-box test (without the `sem.permits` asserts).

File: lib/std/Thread/Semaphore.zig [hunk 2]
```diff
@@ -88,3 +88,16 @@ test "Thread.Semaphore" {
     sem.wait();
     try testing.expect(n == num_threads);
 }
+
+test "Thread.Semaphore - timedWait" {
+    var sem = Semaphore{};
+    try testing.expectEqual(@as(usize, @intCast(0)), sem.permits);
```
Comment (rootbeer) [inline lib/std/Thread/Semaphore.zig:94]: You don't need the casting, just:

    try std.testing.expectEqual(0, sem.permits);

is sufficient (and below too).

**Review by rootbeer — COMMENTED** [review]
New tests look great to me, thanks.  (The whitebox testing is good, makes them much clearer and simpler tests, I think.)  

Hopefully the Zig maintainers will like it too.  :)

The spurious wakeup issue in `timedWait` that @kprotty pointed out should be handled too of course.

**Comment by fkollmann** [comment]
Regarding the spurious wakeup issue, I need you opinion.



## My first approach was working with a deadline

Commit: `96fefe96945c19cfc48f56e5f951608ae4a1af2c`

```zig
pub fn timedWait(sem: *Semaphore, timeout_ns: u64) error{Timeout}!void {
    sem.mutex.lock();
    defer sem.mutex.unlock();

    const deadline = std.time.nanoTimestamp() + @as(i128, @intCast(timeout_ns));
    while (sem.permits == 0) {
        const local_timeout_ns: i64 = @intCast(deadline - std.time.nanoTimestamp());

        try sem.cond.timedWait(&sem.mutex, @bitCast(local_timeout_ns));
    }

    sem.permits -= 1;
    if (sem.permits > 0)
        sem.cond.signal();
}
```

However, I think this is bad, because (a) the function works differently from what the name suggests, and (b) it assumes that the time resolution of `nanoTimestamp()` is nanoseconds, on any platform, which might be incorrect.

Maybe having a `waitDeadline()` could be a way, but I am still not convinced here.

## My second approach was modifying the guarantees:

Commit: `7256e8a0e5e7f1d56b2b12646a348db1bf1a9c9b`

```zig
pub fn timedWait(sem: *Semaphore, timeout_ns: u64) error{Timeout}!void {
    sem.mutex.lock();
    defer sem.mutex.unlock();

    if (sem.permits == 0)
        try sem.cond.timedWait(&sem.mutex, timeout_ns);

    while (sem.permits == 0)
        try sem.cond.timedWait(&sem.mutex, 0); // if another timedWait already

    sem.permits -= 1;
    if (sem.permits > 0)
        sem.cond.signal();
}
```

What this version does, it guarantees, that `timedWait()` will return *before* the timeout, but it could also return *earlier*. (Maybe have an extra error for this like `TimeoutEarly`?)

The basic idea is that `timedWait()` is not being used to wait for a specific amount of time, but to prevent deadlocks in esp. shutdown-scenarios. For this, it is fine if the function returns earlier, if it does not exceeed the specified timeout.

(This of course also would require a documentation comment, which I left out for now.)



Please let me know your thoughts.

File: lib/std/Thread/Semaphore.zig [hunk 3]
```diff
@@ -26,6 +43,18 @@ pub fn wait(sem: *Semaphore) void {
         sem.cond.signal();
 }
 
+pub fn timedWait(sem: *Semaphore, timeout_ns: u64) error{Timeout}!void {
+    sem.mutex.lock();
+    defer sem.mutex.unlock();
+
+    while (sem.permits == 0)
+        try sem.cond.timedWait(&sem.mutex, timeout_ns);
```
[the remark this answers is not in the document]
Comment (fkollmann) [reply lib/std/Thread/Semaphore.zig:51]: see https://github.com/ziglang/zig/pull/18805#issuecomment-1930668241

**Comment by kprotty** [comment]
1. Is the proper approach and there's `std.Thread.Futex.Deadline` which provides this correctly using the system's high precision timer `std.time.Timer` instead of system datetime `std.time.{}timestamp`. Issue is that you'd need to rework Semaphore to use Futex instead of Mutex + Condition.

2. This version is inefficient as, upon spurious wake, it would spin on `cond.timedWait(0)` until a permit is released. This is not equivalent to `cond.wait()` which efficiently puts the thread to sleep.

Stepping back for a bit, I wonder if the stdlib should even provide an `std.Thread.Semaphore` at all? The use case is quite rare, with most situations being better served by either 1) a Mutex + Condition integrated with the shared resource the Semaphore would be protecting or 2) using Futex directly if the resource can be used without locks / with atomics. 

As an example: a ThreadPool should avoid Semaphore given it probably already has a Mutex locked for its work queue so no need to lock another for simply updating a semaphore counter; Replace it with a Condition instead.

**Comment by rootbeer** [comment]
Version 2 seems wrong because an initial spurious wakeup (consider a Semphore with a lot of waiters) would prevent the timeout from kicking in.   I think your version 1 code is missing  the exit when the timeout expires (if `local_timeout_ns` is less than zero).  Expecting `nanoTimestamp` to always return nanoseconds seems reasonable (the granularity of the timer may not operate at that level, but I think that can be covered by the wait being "at least as long as the given timeout").

As @kprotty suggests, the approach used in the Futex timedWait code seems applicable here (it computes a deadline).  (I think its fine to just be "inspired" by that code at this point.)

I think the Semaphore API can be simpler, in some cases, for callers to use than a mutex+condvar, so it seems reasonable to support in the standard library, but I don't have any concrete use cases for it.

**Comment by fkollmann** [comment]
Thanks for your feedback @kprotty @rootbeer.

I did change the implementation to be based on `std.time.Timer` and think it has a much cleaner design. Please let me know your thoughts.

**Comment by fkollmann** [comment]
Thanks for your feedback @kprotty @rootbeer.

I did change the implementation to be based on `std.time.Timer` and think it has a much cleaner design. Please let me know your thoughts.

**Comment by rootbeer** [comment]
LGTM.

**Comment by fkollmann** [comment]
@kprotty please let me know what you think. If it's fine by you, please update the branch and approve the workflow.

**Pull request merged**